content/blog/2016/2016-10-11-stage-lock-milestone.adoc

```diff
@@ -0,0 +1,125 @@
+
+* `lock` can be used to constrain just a sub-section of a `stage`
+
+[source, groovy]
+----
+stage("Build"){
+  doSomething()
+    lock("myResource"){
+    echo "locked build"
+  }
+}
+----
+
+* `lock` can be used to wrap multiple stages into a single concurrency unit
+
+[source, groovy]
+----
+lock("myResource"){
+  stage("Build"){
+    echo "Building"
+  }
+  stage("Test"){
+    echo "Testing"
+  }
+}
+----
+
+=== milestone
+
+The `milestone` step is the last piece of the puzzle to replace functionality originally intended for `stage` and adds even more control for handling concurrent builds of a job. The `lock` step limits the number of builds running concurrently in a section of your Pipeline while the `milestone` step ensures that older builds of a job will not overwrite or clobber a newer build.
+
+Concurrent builds of the same job do not always run at the same rate. Depending on the network, the node used, compilation times, test times, etc. it is always possible for a newer build to complete faster than an older build.
+
+* Build A is triggered
+* Build B is triggered
+* Build B builds faster than Build A and enters the Test stage sooner.
+
+Rather than allowing Build A to continue and possibly overwrite the newer artifact produced in Build B you can use the `milestone` step to abort Build A:
+
+[source, groovy]
+----
+stage("Build"){
+  milestone
+  echo "Building"
+}
+stage("Test"){
+  milestone
+  echo "Testing"
+}
+----
+
+When using the `input` step or the `lock` step it becomes much easier to build up a backlog of concurrent builds either waiting for user input or waiting for a resource to become free. The `milestone` step will automatically prune all older jobs that are waiting at these junctions.
+
+[source, groovy]
+----
+milestone
+input message: "Proceed?"
+milestone
+----
+
+Bookending an `input` step like this allows you to select a specific build to proceed and automatically abort all antecedent builds.
+
+[source, groovy]
+----
+milestone
+lock(resource: 'myResource', inversePrecedence: true){
+  echo "locked step"
+  milestone
+}
+----
+
+Similarly a pair of `milestone` steps used with a `lock` will discard all old builds waiting for a shared resource. In this example, the inversePrecedence argument instructs the `lock` to begin newest build in the queue first ensuring that the most recent code takes precedence.
+
+=== Putting it all together
+
+The link:https://github.com/jenkinsci/workflow-aggregator-plugin/blob/master/demo/repo/Jenkinsfile[Jenkinsfile] provided with the link:https://github.com/jenkinsci/workflow-aggregator-plugin/tree/master/demo[Docker image for demonstrating Pipeline] utilizes all of the steps described here as well as several others. Please refer to it for a more complete example.
```

Comment:
Would it be possible to show all three steps in one example with annotation about happens before you refer out to an external location?   As a reader I'd like to see "putting it all together" before I click out to explore further.